### ISSUE TITLE
LLVM generates bad code on newer architercures

### ISSUE THREAD

**Issue by pabloferz** [JuliaLang/julia#19976]
Didn't find an issue for this. Cross-reference: https://github.com/JuliaLang/julia/pull/19259#issuecomment-270640272

I believe some of the performance "noisy" regressions seen lately are caused by this.

On my machine (Haswell) the [pisum()](https://github.com/JuliaLang/julia/blob/30bf89f3d8e564b588b8e48993e92a551b384f2c/test/perf/micro/perf.jl#L84) benchmark on master gives

```julia
julia> @benchmark pisum()
BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     33.362 ms (0.00% GC)
  median time:      33.445 ms (0.00% GC)
  mean time:        33.532 ms (0.00% GC)
  maximum time:     34.761 ms (0.00% GC)
  --------------
  samples:          150
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

while on 0.5 I get

```julia
julia> @benchmark pisum()
BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     7.548 ms (0.00% GC)
  median time:      7.571 ms (0.00% GC)
  mean time:        7.611 ms (0.00% GC)
  maximum time:     8.404 ms (0.00% GC)
  --------------
  samples:          657
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

Even when both show practically the same LLVM code.

cc @yuyichao

**Comment by vchuravy** [comment]
A direct way of seeing the effect on the same version of Julia is:

```
julia -C core2 --precompiled=no pisum.jl
@benchmark(pisum()) = BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     19.512 ms (0.00% GC)
  median time:      19.795 ms (0.00% GC)
  mean time:        19.797 ms (0.00% GC)
  maximum time:     21.656 ms (0.00% GC)
  --------------
  samples:          253
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

vs:

```
julia -C haswell --precompiled=no pisum.jl
@benchmark(pisum()) = BenchmarkTools.Trial: 
  memory estimate:  0.00 bytes
  allocs estimate:  0
  --------------
  minimum time:     37.291 ms (0.00% GC)
  median time:      37.888 ms (0.00% GC)
  mean time:        37.931 ms (0.00% GC)
  maximum time:     39.078 ms (0.00% GC)
  --------------
  samples:          132
  evals/sample:     1
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

**Comment by yuyichao** [comment]
It's not a regression, I've checked that both LLVM 3.3 and 3.7.1 gives the same bad code when compiling with avx enabled.

This might have been fixed on LLVM trunk right after 3.9 is branched and I'm trying to backport the patch.
